src/lib/pubkey/sphincsplus/sphincsplus_common/sphincsplus.cpp

```diff
@@ -19,14 +20,53 @@
 #include <botan/internal/sp_xmss.h>
 #include <botan/internal/stl_util.h>
 
-#if !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2) and !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE)
+#include <utility>
+
+#if !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHA2) and !defined(BOTAN_HAS_SPHINCS_PLUS_WITH_SHAKE) and \
+   !defined(BOTAN_HAS_SLH_DSA_WITH_SHA2) and !defined(BOTAN_HAS_SLH_DSA_WITH_SHAKE)
 static_assert(
    false,
-   "botan module 'sphincsplus_common' is useful only when enabling at least 'sphincsplus_sha2' or 'sphincsplus_shake'");
+   "botan module 'sphincsplus_common' is useful only when enabling at least 'sphincsplus_sha2', 'sphincsplus_shake', 'slh_dsa_sha2', or 'slh_dsa_shake'");
 #endif
 
 namespace Botan {
 
+namespace {
+// FIPS 205, Algorithm 22, line 8 and Algroithm 23, line 24
+std::vector<uint8_t> prepare_message(std::vector<uint8_t> msg,
+                                     const Sphincs_Parameters& params,
+                                     std::span<const uint8_t> context) {
+   if(params.is_slh_dsa()) {
+      const size_t OID_BYTE_SIZE = 11;
+      // prefix: input mode byte + |ctx| + ctx + OID(for pre-hash mode only)
+      std::vector<uint8_t> prefix(1 + 1 + context.size() +
+                                  (params.input_mode() == SlhDsaInputMode::PreHash ? OID_BYTE_SIZE : 0));
+      BufferStuffer prefix_stuffer(prefix);
+      uint8_t input_mode_byte = [&]() -> uint8_t {
+         switch(params.input_mode()) {
+            case SlhDsaInputMode::Pure:
+               return 0x00;
+            case SlhDsaInputMode::PreHash:
+               return 0x01;
+         }
+         BOTAN_ASSERT_UNREACHABLE();
+      }();
+      prefix_stuffer.append(input_mode_byte);
+      prefix_stuffer.append(checked_cast_to<uint8_t>(context.size()));
+      prefix_stuffer.append(context);
+      if(params.input_mode() == SlhDsaInputMode::PreHash) {
+         throw Not_Implemented("PreHash mode not implemented yet");
+      }
+      BOTAN_ASSERT_NOMSG(prefix_stuffer.full());
+
+      msg.insert(msg.begin(), prefix.begin(), prefix.end());
+   }
+   // else: SPHINCS+ Round 3.1 uses the message without any prefix
```

Comment:
The problem is that the logic for prepare_message differs in the SLH-DSA or SPHINCS+ dimension, while the hash function is in the SHA2 and SHAKE dimension. These two dimensions are orthogonal. Since the logic for the hashes is shared between SLH-DSA and SPHINCS+, I currently only need two derived classes (one for SHA2 and one for SHAKE). Adding an additional layer of module and class hierarchy for the prepare_message function (15 lines of code) alone is overkill, in my opinion.

Reply:
Alternatively, you could `#ifdef` the implementations depending on whether `BOTAN_HAS_SLH_DSA` and/or `BOTAN_HAS_SPHINCSPLUS` and add a default error case if necessary.

Reply:
I'll try that 👍